### PR TITLE
Introduced CreateDatumCast & CreateCast API

### PR DESCRIPTION
**Pull request by karthijrk**
In this PR, I introduced `CreateCppTypeToDatumCast `, `CreateDatumToCppTypeCast` and `CreateCast` to handle casting in IR. Added necessary unit test for this api.

Also refactored existing codegeneration code for expression evaluation as well as PG Function generator. Refactor includes usage of above introduced new api and elimination of duplicate codes.

Note : Template specialization for CreateCast is not complete yet. This PR supports only any integer to integer type or any floating point conversion. Those are the use case we have right now. We can enhance as we encountered different casting use case.

@foyzur @hardikar Please take a look when you get chance.

### CODE REVIEW

File: src/backend/codegen/include/codegen/pg_arith_func_generator.h [hunk 1]
```diff
@@ -27,6 +27,35 @@ namespace gpcodegen {
  *  @{
  */
 
+// ArithOpOverFlowErrorMsg has various template specializations to
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/pg_arith_func_generator.h]: Just some thoughts.. 
(1) Do you think it's easier to have overloaded methods instead of template classes? 
(2) Do you think we should put these in some other namespace, since they're not intended for use in general in gpcodegen, am I right?

File: src/backend/codegen/include/codegen/pg_arith_func_generator.h [hunk 2]
```diff
@@ -37,6 +66,11 @@ namespace gpcodegen {
  **/
 template <typename rtype, typename Arg0, typename Arg1>
 class PGArithFuncGenerator {
+  template <typename CppType>
+  using CGArithOpOverFlowTemplateFunc = llvm::Value* (GpCodegenUtils::*)(
+      llvm::Value* arg0, llvm::Value* arg1);
+  using CGArithOpOverFlowFunc = CGArithOpOverFlowTemplateFunc<rtype>;
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/pg_arith_func_generator.h]: Is there really any value is carrying the OverFlow in all these names? I mean are there non-overflow functions at all?

File: src/backend/codegen/include/codegen/pg_arith_func_generator.h [hunk 3]
```diff
@@ -95,130 +151,81 @@ class PGArithFuncGenerator {
...
+        irb->CreateExtractValue(llvm_arith_output, 1);
 
     irb->CreateCondBr(llvm_overflow_flag,
                       llvm_overflow_block,
                       llvm_non_overflow_block);
 
     irb->SetInsertPoint(llvm_overflow_block);
-    // TODO(krajaraman) Elog::ERROR after ElogWrapper integrated.
+    codegen_utils->CreateElog(
+        ERROR, "%s", llvm_error_msg);
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/pg_arith_func_generator.h]: This is the only place you're using llvm_error_msg. Do you think it is possible to getconstant here, since this function is templated by rtype. That way we can avoid another member that's passed around :)

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 4]
```diff
@@ -106,6 +106,13 @@ class CodegenUtils {
   }
 
   /**
+   * @return LLVMContext
+   **/
+  llvm::LLVMContext* context() {
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: Any reason this needs to be public (and not protected?)

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 5]
```diff
@@ -1237,50 +1228,46 @@ std::is_integral<UnsignedIntType>::value
   static llvm::Value* CreateAddOverflow(CodegenUtils* generator,
                                         llvm::Value* arg0,
                                         llvm::Value* arg1) {
-    Checker(arg0, arg1);
-    llvm::Value* casted_arg0 = generator->CreateCast<UnsignedIntType>(arg0);
-    llvm::Value* casted_arg1 = generator->CreateCast<UnsignedIntType>(arg1);
-
+    Checker(generator->GetType<UnsignedIntType>(), arg0, arg1);
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: This may be out of scope of this PR - but can we rename this function? :P Checker is (a) not a verb - so it feels like a struct/class to me; and (b) too vague a name for what it's doing.

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 6]
```diff
@@ -1437,6 +1409,160 @@ llvm::Value* CodegenUtils::CreateMulOverflow(llvm::Value* arg0,
...
+class CastMaker<
+    IntegerType,
+    UnsignedIntType,
+    typename std::enable_if<
+        std::is_integral<IntegerType>::value
+        && std::is_integral<UnsignedIntType>::value
+        && std::is_unsigned<UnsignedIntType>::value>::type> {
+ public:
+  static llvm::Value* CreateCast(llvm::Value* value,
+                                 llvm::Type* llvm_dest_type,
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: Why do we need this, given knowledge of the DestType? Can you not figure this out by using codegen_utils->GetType<DestType>() ?

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 7]
```diff
@@ -1437,6 +1409,160 @@ llvm::Value* CodegenUtils::CreateMulOverflow(llvm::Value* arg0,
...
+template <typename DestType,
+          typename SrcType,
+          typename Enable = void>
+class CastMaker {
+};
+
+// Partial specialization for unsigned int to signed / unsigned integer.
+template <typename IntegerType, typename UnsignedIntType>
+class CastMaker<
+    IntegerType,
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: consistent naming : SignedIntegerType to UnsignedIntegerType

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 8]
```diff
@@ -1437,6 +1409,160 @@ llvm::Value* CodegenUtils::CreateMulOverflow(llvm::Value* arg0,
...
+               llvm::Type* llvm_dest_type) {
+    assert(nullptr != value);
+    llvm::Type* llvm_src_type = value->getType();
+    assert(nullptr != llvm_src_type);
+    assert(llvm_src_type->isIntegerTy());
+    assert(llvm_dest_type->isIntegerTy());
+  }
+};
+
+// Partial specialization for signed int to signed / unsigned integer.
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: Update comment - this looks like it's for signed int -> signed  int (which also looks wierd).
In cpp land, which is what IntegerType is in, it's either signed or unsigned - and can't be both or neither AFAIK.

File: src/backend/codegen/include/codegen/pg_arith_func_generator.h (hunk 1 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/pg_arith_func_generator.h]: I agree, we can put in different namespace something like ArithOp_details. I think in future we might need to put more static method in this class for different use case. For that having template class might be more useful.

File: src/backend/codegen/include/codegen/pg_arith_func_generator.h (hunk 2 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/pg_arith_func_generator.h]: Actually you are right. This means to hold pointer to ArithOp function. CGArithOpFunc & CGArithOpTemplateFunc works ?

File: src/backend/codegen/include/codegen/pg_arith_func_generator.h (hunk 3 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/pg_arith_func_generator.h]: Do you mean call GetContant directly here on error message string ? For date, it doesn't hold true unfortunately. We need to pass specific message

File: src/backend/codegen/include/codegen/utils/codegen_utils.h (hunk 4 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: Agree. Lets move it to protected.

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 9]
```diff
@@ -1437,6 +1409,160 @@ llvm::Value* CodegenUtils::CreateMulOverflow(llvm::Value* arg0,
...
+    float,
+    FloatingPoint,
+    typename std::enable_if<
+            std::is_floating_point<FloatingPoint>::value>::type> {
+ public:
+  static llvm::Value* CreateCast(llvm::Value* value,
+                                 llvm::Type* llvm_dest_type,
+                                 CodegenUtils* codegen_utils) {
+    Checker(value, llvm_dest_type);
+    if (value->getType()->isFloatTy()) { return value; }
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: There is something I don't understand here. What is the benefit of all these template specializations if we still have to end up checking the value's type?

File: src/backend/codegen/include/codegen/utils/codegen_utils.h (hunk 6 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: May be. But I have to repeat same GetType on all specialization. Found more convenient to get as an argument

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 10]
```diff
@@ -1437,6 +1409,160 @@ llvm::Value* CodegenUtils::CreateMulOverflow(llvm::Value* arg0,
...
+    assert(llvm_dest_type->isDoubleTy());
+  }
+};
+
+}  // namespace codegen_utils_detail
+
+template <typename DestType, typename SrcType>
+llvm::Value* CodegenUtils::CreateCast(llvm::Value* value) {
+  return codegen_utils_detail::CastMaker<DestType, SrcType>::CreateCast(
+      value, this->GetType<DestType>(), this);
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: Looks like you don't need to compute GetType<DestType>() here, and we can reduce the number of parameters!

File: src/backend/codegen/include/codegen/utils/codegen_utils.h (hunk 8 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: Not sure if I fully understood the comment.
But e.g. for signed int to signed int is, 
```
int32_t a = 23;
int64_t b = (int32_t)a;
```

File: src/backend/codegen/include/codegen/utils/codegen_utils.h (hunk 10 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: or I don't need to have the same code all the place! :)

File: src/backend/codegen/include/codegen/utils/gp_codegen_utils.h [hunk 11]
```diff
@@ -101,6 +102,72 @@ class GpCodegenUtils : public CodegenUtils {
...
+   * @param value     LLVM Value on which casting has to be done.
+   *
+   * @return LLVM Value that casted to given Cpp type.
+   **/
+  template <typename CppType>
+  llvm::Value* CreateDatumToCppTypeCast(llvm::Value* value) {
+    assert(nullptr != value);
+    llvm::Type* llvm_dest_type = GetType<CppType>();
+    // If it is a pointer, do direct pointer cast
+    if (llvm_dest_type->isPointerTy()) {
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/gp_codegen_utils.h]: Same argument as above. I don't understand why this function is templatized, since the logic is a bunch of if-checks.

File: src/backend/codegen/utils/gp_codegen_utils.cc [hunk 12]
```diff
@@ -19,6 +19,50 @@
 #include "codegen/utils/gp_codegen_utils.h"
 
 namespace gpcodegen {
+
+llvm::Value* GpCodegenUtils::CreateCppTypeToDatumCast(
```
Comment (hardikar) [inline src/backend/codegen/utils/gp_codegen_utils.cc]: I think given the size of CreateDatumToCppTypeCast, we should also bring that down here.

File: src/backend/codegen/include/codegen/utils/codegen_utils.h (hunk 9 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: Its a trade off. Given we have src_type and dest_type I don't like to write too many specialization for different combination.

File: src/backend/codegen/include/codegen/utils/gp_codegen_utils.h (hunk 11 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/utils/gp_codegen_utils.h]: Thats a good idea. This will be a partial template specialization for pointers which is not supported for function. I have to get a class similar to what we did for GetType. If you guys think  have this if is too complex / regression we can discuss on having partial specialization class.

**Comment by hardikar** [comment]
+1

**Comment by hardikar** [comment]
FYI - I see some warnings when building:
```
/Users/shardikar/workspace/codegen/gpdb/src/backend/codegen/include/codegen/utils/codegen_utils.h:1456:17: warning: unused variable 'llvm_src_type' [-Wunused-variable]
    llvm::Type* llvm_src_type = value->getType();
                ^
/Users/shardikar/workspace/codegen/gpdb/src/backend/codegen/include/codegen/utils/codegen_utils.h:1486:17: warning: unused variable 'llvm_src_type' [-Wunused-variable]
    llvm::Type* llvm_src_type = value->getType();
                ^
/Users/shardikar/workspace/codegen/gpdb/src/backend/codegen/include/codegen/utils/codegen_utils.h:1514:19: warning: unused variable 'llvm_src_type' [-Wunused-variable]
      llvm::Type* llvm_src_type = value->getType();
                  ^
/Users/shardikar/workspace/codegen/gpdb/src/backend/codegen/include/codegen/utils/codegen_utils.h:1543:17: warning: unused variable 'llvm_src_type' [-Wunused-variable]
    llvm::Type* llvm_src_type = value->getType();
```

**Comment by karthijrk** [comment]
I used the llvm_src_type in assert. Is it because codegen built without assert and it gives warning ?

File: src/backend/codegen/utils/gp_codegen_utils.cc [hunk 13]
```diff
@@ -19,6 +19,51 @@
...
+    llvm::Value* value, bool is_src_unsigned) {
+  assert(nullptr != value);
+
+  llvm::Type* llvm_dest_type = GetType<Datum>();
+  unsigned dest_size = llvm_dest_type->getScalarSizeInBits();
+  assert(sizeof(Datum) << 3 == dest_size);
+
+  llvm::Type* llvm_src_type = value->getType();
+  unsigned src_size = llvm_src_type->getScalarSizeInBits();
+  if (llvm_src_type->isPointerTy()) {
```
Comment (foyzur) [inline src/backend/codegen/utils/gp_codegen_utils.cc]: Comment?

File: src/backend/codegen/op_expr_tree_generator.cc [hunk 14]
```diff
@@ -65,19 +65,19 @@ void OpExprTreeGenerator::InitializeSupportedFunction() {
       new PGGenericFuncGenerator<float8, float8>(
           216,
           "float8mul",
-          &PGArithFuncGenerator<float8, float8, int64_t>::MulWithOverflow));
+          &PGArithFuncGenerator<float8, float8, float8>::MulWithOverflow));
```
Comment (foyzur) [inline src/backend/codegen/op_expr_tree_generator.cc]: Is this relevant to casting?

File: src/backend/codegen/op_expr_tree_generator.cc [hunk 15]
```diff
@@ -181,9 +181,13 @@ bool OpExprTreeGenerator::GenerateCode(GpCodegenUtils* codegen_utils,
     }
     llvm_arguments.push_back(llvm_arg);
   }
-  return itr->second->GenerateCode(codegen_utils,
+  llvm::Value* llvm_op_value = nullptr;
+  bool retval = itr->second->GenerateCode(codegen_utils,
```
Comment (foyzur) [inline src/backend/codegen/op_expr_tree_generator.cc]: Can we comment on itr->second and save that in a variable and then use it?

File: src/backend/codegen/pg_date_func_generator.cc [hunk 16]
```diff
@@ -33,21 +34,15 @@ bool PGDateFuncGenerator::DateLETimestamp(
...
 
-  llvm::Value* llvm_arg0_DateADT = codegen_utils->
-      CreateCast<DateADT>(llvm_args[0]);
-  /* TODO: Use CreateCast*/
-  llvm::Value* llvm_arg1_Timestamp = irb->CreateSExt(
-      llvm_args[1], codegen_utils->GetType<Timestamp>());
-
   llvm::Value* llvm_arg0_Timestamp = GenerateDate2Timestamp(
-      codegen_utils, llvm_main_func, llvm_arg0_DateADT, llvm_error_block);
+      codegen_utils, llvm_main_func, llvm_args[0], llvm_error_block);
```
Comment (foyzur) [inline src/backend/codegen/pg_date_func_generator.cc]: Can we put an in-place comment for unnamed parameters (e.g., llvm_args[0] /* ... */)

File: src/backend/codegen/utils/gp_codegen_utils.cc [hunk 17]
```diff
@@ -19,6 +19,51 @@
...
+  llvm::Value* llvm_casted_value = llvm_int_value;
+  if (src_size < dest_size) {
+    // If a given src type is integer but not a bool and unsigned flag is set to
+    // false, then do signed extension
+    if (llvm_src_type->isIntegerTy() &&
+        src_size > 1 &&
+        !is_src_unsigned) {
+      llvm_casted_value = ir_builder()->CreateSExt(llvm_int_value,
+                                                   llvm_dest_type);
+    } else {
```
Comment (foyzur) [inline src/backend/codegen/utils/gp_codegen_utils.cc]: Else is typically a complete negation of single if. Are you sure you have that semantics here?

File: src/backend/codegen/exec_eval_expr_codegen.cc [hunk 18]
```diff
@@ -150,7 +150,7 @@ bool ExecEvalExprCodegen::GenerateExecEvalExpr(
     return false;
   }
 
-  llvm::Value* llvm_ret_value = codegen_utils->CreateCast<int64_t>(value);
+  llvm::Value* llvm_ret_value = codegen_utils->CreateCppTypeToDatumCast(value);
```
Comment (foyzur) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: How do you differentiate between signed and unsigned cpp type?

**Comment by foyzur** [comment]
LGTM with few comments.

File: src/backend/codegen/op_expr_tree_generator.cc (hunk 14 above)
Comment (karthijrk) [inline src/backend/codegen/op_expr_tree_generator.cc]: Yes, I moved the responsibility of converting float to datum to ExprTreeGenerator.

File: src/backend/codegen/utils/gp_codegen_utils.cc (hunk 17 above)
Comment (karthijrk) [inline src/backend/codegen/utils/gp_codegen_utils.cc]: Yes, I add some comments on else part

File: src/backend/codegen/exec_eval_expr_codegen.cc (hunk 18 above)
Comment (karthijrk) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: I don't see greenplum / postgres allow unsigned data type. By default CreateCppTypeToDatumCast will treat as signed integer. 
http://gpdb.docs.pivotal.io/4370/ref_guide/data_types.html

In case, if we need to allow unsigned, then each ExprTree need to return as Datum instead of actual type. Each Exprtree generator knows what type they are returning and hence when they convert to Datum they can specify whether they are signed or unsigned.
